src/app/clusters/im-compatibility/compatibility-command-handlers.cpp

```diff
@@ -0,0 +1,841 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+// THIS FILE IS GENERATED BY ZAP
+
+#include <cinttypes>
+#include <cstdint>
+
+#include <gen/af-structs.h>
+#include <gen/call-command-handler.h>
+#include <gen/callback.h>
+#include <gen/cluster-id.h>
+#include <gen/command-id.h>
+#include <util.h>
+
+#include <app/InteractionModelEngine.h>
+#include <gen/IMClusterCommandHandler.h>
+
+#include "ember-compatibility-functions.h"
+
+namespace chip {
+namespace app {
+
+// Cluster specific command parsing
+// TODO: Take care of internal status of ember library.
+
+namespace clusters {
+
+namespace BarrierControl {
+#if IM_HAVE_BARRIER_CONTROL_CLUSTER_BARRIER_CONTROL_GO_TO_PERCENT_COMMAND
+void OnBarrierControlGoToPercentCommandCallback(app::Command * command, EndpointId ZCLEndpointId, uint8_t percentOpen)
+{
+    Compatibility::SetupEmberAfObjects(command, ZCL_BARRIER_CONTROL_CLUSTER_ID, ZCL_BARRIER_CONTROL_GO_TO_PERCENT_COMMAND_ID,
+                                       ZCLEndpointId);
+    emberAfBarrierControlClusterBarrierControlGoToPercentCallback(percentOpen);
+    Compatibility::ResetEmberAfObjects();
```

Comment:
Why do you need `src/app/clusters/im-compatibility/compatibility-command-handlers.cpp` ? Can't you directly call `Compatibility::SetupEmberAfObjects(command, ZCL_BARRIER_CONTROL_CLUSTER_ID, commandId, endpointId);` at the beginning of `Clusters::BarrierControl::DispatchServerCommand`, `emberAf{{ClusterName}}Cluster{{CommandName}}Callback` in the switch/case, and `Compatibility::ResetEmberAfObjects();` at the end ?

Reply:
I'd like to make the Command object available for command handlers, and the future commands can be written in the IM style. Anyway, I think it is a good idea that we can call `Compatibility::SetupEmberAfObjects` and `Compatibility::ResetEmberAfObjects`  from the dispatcher.

Reply:
Then it will be better to fix the current commands handlers to have a new parameter for it in a separate PR.
Please remove `src/app/clusters/im-compatibility`, it introduce a new generated folder in `src/` and creates a level of indirection that does not seems really needed.

Reply:
Issue: https://github.com/project-chip/connectedhomeip/issues/5098